File: leetcode/rust/problems/src/hard/_23_merge_k_sorted_lists.rs
// Leetcode Problem 23

use std::collections::BinaryHeap;

use utils::*;

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn merge_k_lists(lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
        let mut solution = None;

        let mut queue: BinaryHeap<i32> = BinaryHeap::new();

        for mut list in lists {
            while let Some(item) = list {
                queue.push(item.val);
                list = item.next;
            }
        }

        while let Some(value) = queue.pop() {
            solution = Some(Box::new(ListNode{val: value, next: solution}));
        }

        solution
    }
}


#[test]
fn test_merge_k_lists() {
    let list1 = list!(1, 4, 5);
    let list2 = list!(1, 3, 4);
    let list3 = list!(2, 6);

    let sol1 = list!(1, 1, 2, 3, 4, 4, 5, 6);

    assert_eq!(Solution::merge_k_lists(vec![]), list!());
    assert_eq!(Solution::merge_k_lists(vec![list!()]), list!());
    assert_eq!(Solution::merge_k_lists(vec![list1, list2, list3]), sol1);
}

File: leetcode/c++/hard/_23_merge_k_sorted_lists.cpp
// Leetcode Problem 23

#include <gtest/gtest.h>
#include <vector>
#include <queue>
#include "LinkedList.h"
#include "BinaryMinHeap.h"

using namespace std;

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto comp = [](ListNode* a, ListNode* b) {
            return a->val > b->val;
        };

        priority_queue<ListNode*, vector<ListNode*>, decltype(comp)> minHeap(comp);
    
        for (auto head : lists) {
            if (head) {
                minHeap.push(head);
            }
        }

        ListNode* head = new ListNode();
        ListNode* tempHead = head;

        while (!minHeap.empty()) {
            ListNode* temp = minHeap.top();
            minHeap.pop();

            head->next = temp;
            head = head->next;

            if (temp && temp->next) {
                minHeap.push(temp->next);
            }
        }

        return tempHead->next;
    }   

    /**
     * Solution function using own min heap implementation
     */
    ListNode* mergeKLists_customHeap(vector<ListNode*>& lists) {
        if (lists.empty()) {
            return nullptr;
        }

        MinHeap<int> heap;

        for (int i = 0; i < lists.size(); i++) {
            while (lists[i] != nullptr) {
                heap.insert(lists[i]->val);
                lists[i] = lists[i]->next;
            }
        }

        if (heap.size() == 0) {
            return nullptr;
        }

        ListNode* head = new ListNode(heap.extractMin());
        ListNode* temp = head;

        while (!heap.isEmpty()) {
            head->next = new ListNode(heap.extractMin());
            head = head->next;
        }
        
        return temp;
    }
};


class TestMergeKLists : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestMergeKLists, SimpleMergeCase) {
    vector<ListNode*> lists = {
        createLinkedList({1, 4, 5}),
        createLinkedList({1, 3, 4}),
        createLinkedList({2, 6})
    };

    ListNode* sol = createLinkedList({1, 1, 2, 3, 4, 4, 5, 6});

    EXPECT_TRUE(compareLists(sol, solution.mergeKLists(lists)));
}

TEST_F(TestMergeKLists, EmptyLists) {
    vector<ListNode*> lists = {};

    EXPECT_TRUE(compareLists(nullptr, solution.mergeKLists(lists)));
}

TEST_F(TestMergeKLists, EmptyList) {
    vector<ListNode*> lists = {{}};

    EXPECT_TRUE(compareLists(nullptr, solution.mergeKLists(lists)));   
}